dev/tests/static/testsuite/Magento/Test/Legacy/Magento/App/Action/AbstractActionTest.php

```diff
@@ -0,0 +1,160 @@
+    {
+        $files = $this->getTestFiles();
+        $found = [];
+
+        foreach ($files as $file) {
+            $class = $this->classNameExtractor->getNameWithNamespace(file_get_contents($file[0]));
+
+            if ($class) {
+                try {
+                    $classReflection = new  ReflectionClass($class);
+                    if ($classReflection->isSubclassOf(AbstractAction::class)) {
+                        $found[] = $class;
+                    }
+                } catch (Exception $exception) {
+                    $this->addWarning('Skipped due to exception: ' . $class);
+                }
+            }
+        }
+
+        $this->assertEmpty(
+            $found,
+            "The following new controller(s) extend " . AbstractAction::class . "\r\n"
+            . "All new controller classes must implement " . ActionInterface::class . " instead.\r\n"
+            . print_r($found, true)
+        );
+    }
+
+    /**
+     * Provide files for test.
+     *
+     * @return array
+     */
+    private function getTestFiles(): array
+    {
+        $phpFiles = self::getAddedFilesList(self::getChangedFilesBaseDir());
+
+        $phpFiles = Files::composeDataSets($phpFiles);
+        $fileTypes = Files::INCLUDE_APP_CODE | Files::INCLUDE_LIBS | Files::AS_DATA_SET;
+        return array_intersect_key($phpFiles, $this->fileUtilities->getPhpFiles($fileTypes));
+    }
+
+    /**
+     * Provide list of new files.
+     *
+     * @param $changedFilesBaseDir
+     *
+     * @return string[]
+     */
+    private static function getAddedFilesList($changedFilesBaseDir)
+    {
+        return self::getFilesFromListFile(
+            $changedFilesBaseDir,
+            'changed_files*.added.*',
+            function () {
+                // if no list files, probably, this is the dev environment
+                // phpcs:ignore Generic.PHP.NoSilencedErrors,Magento2.Security.InsecureFunction
+                @exec('git diff --cached --name-only --diff-filter=A', $addedFiles);
+                return $addedFiles;
+            }
+        );
+    }
+
+    /**
+     * Read files from generated lists.
+     *
+     * @param string $listsBaseDir
+     * @param string $listFilePattern
+     * @param callable $noListCallback
+     * @return string[]
+     */
+    private static function getFilesFromListFile(
+        string $listsBaseDir,
+        string $listFilePattern,
+        callable $noListCallback
+    ): array {
+        $filesDefinedInList = [];
+
+        $listFiles = glob($listsBaseDir . '/_files/' . $listFilePattern);
+        if (!empty($listFiles)) {
+            foreach ($listFiles as $listFile) {
+                // phpcs:ignore Magento2.Performance.ForeachArrayMerge.ForeachArrayMerge
+                $filesDefinedInList = array_merge(
+                    $filesDefinedInList,
+                    file($listFile, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES)
+                );
+            }
```

Comment:
```suggestion
            foreach ($listFiles as $listFile) {
                $filesDefinedInList[] = file($listFile, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
            }
            if (count($filesDefinedInList) > 0) {
                 $filesDefinedInList = array_merge(...$filesDefinedInList);
            }
```

Reply:
@kandy maybe something like that?
```suggestion
            foreach ($listFiles as $listFile) {
                $filesDefinedInList[] = file($listFile, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
            }

           $filesDefinedInList = array_merge([], ...$filesDefinedInList);
```